controllers/rabbitmq/rabbitmq_controller.go

```diff
@@ -0,0 +1,494 @@
+	}
+
+	IPv6Enabled, err := ocp.FirstClusterNetworkIsIPv6(ctx, helper)
+	if err != nil {
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			condition.ServiceConfigReadyCondition,
+			condition.ErrorReason,
+			condition.SeverityWarning,
+			condition.ServiceConfigReadyErrorMessage,
+			err.Error()))
+		return ctrl.Result{}, fmt.Errorf("error getting cluster IPv6 config: %w", err)
+	}
+
+	fipsEnabled, err := ocp.IsFipsCluster(ctx, helper)
+	if err != nil {
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			condition.ServiceConfigReadyCondition,
+			condition.ErrorReason,
+			condition.SeverityWarning,
+			condition.ServiceConfigReadyErrorMessage,
+			err.Error()))
+		return ctrl.Result{}, fmt.Errorf("error getting cluster FIPS config: %w", err)
+	}
+
+	rabbitmqCluster := &rabbitmqv2.RabbitmqCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      instance.Name,
+			Namespace: instance.Namespace,
+		},
+	}
+
+	err = instance.Spec.MarshalInto(&rabbitmqCluster.Spec)
+	if err != nil {
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			condition.ServiceConfigReadyCondition,
+			condition.ErrorReason,
+			condition.SeverityWarning,
+			condition.ServiceConfigReadyErrorMessage,
+			err.Error()))
+		return ctrl.Result{}, fmt.Errorf("error creating RabbitmqCluster Spec: %w", err)
+	}
+
+	instance.Status.Conditions.MarkTrue(condition.ServiceConfigReadyCondition, condition.ServiceConfigReadyMessage)
+
+	//
+	// Handle Topology
+	//
+	topology, err := topologyv1.EnsureServiceTopology(
+		ctx,
+		helper,
+		instance.Spec.TopologyRef,
+		instance.GetLastAppliedTopologyRef(),
+		instance.Name,
+		labels.GetLabelSelector(
+			map[string]string{
+				labels.K8sAppName:      instance.Name,
+				labels.K8sAppComponent: "rabbitmq",
+				labels.K8sAppPartOf:    "rabbitmq",
+			},
+		),
+	)
+	if err != nil {
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			condition.TopologyReadyCondition,
+			condition.ErrorReason,
+			condition.SeverityWarning,
+			condition.TopologyReadyErrorMessage,
+			err.Error()))
+		return ctrl.Result{}, fmt.Errorf("waiting for Topology requirements: %w", err)
+	}
+
+	// If TopologyRef is present and ensureHeatTopology returned a valid
+	// topology object, set .Status.LastAppliedTopology to the referenced one
+	// and mark the condition as true
+	if instance.Spec.TopologyRef != nil {
+		// update the Status with the last retrieved TopologyRef
+		instance.Status.LastAppliedTopology = instance.Spec.TopologyRef
+		// update the TopologyRef associated condition
+		instance.Status.Conditions.MarkTrue(condition.TopologyReadyCondition, condition.TopologyReadyMessage)
+	} else {
+		// remove LastAppliedTopology from the .Status
+		instance.Status.LastAppliedTopology = nil
+	}
+
+	rabbitmqImplCluster := impl.NewRabbitMqCluster(rabbitmq.ConfigureCluster(rabbitmqCluster, IPv6Enabled, fipsEnabled, topology, instance.Spec.NodeSelector), 5)
```

Comment:
We should pass to the `NewRabbitMqCluster` the `serviceLabels` used as `defaultLabelSelector` for the topology on L309, and update the `NewRabbitMqCluster` signature to accept them (and pass down to the `statefulSet .Spec.Selector`).

Reply:
no, rabbitmqcluster sets the labels, do we just set defaultLabelSelector to the same values it uses

Reply:
we could use the override to change the default labels of the rabbimqcluster, BUT this will break existing deployments. so I think we can not do it.

Reply:
ok, so, no labels passing to the underlying statefulSet, +1 to not break any existing deployment. In this case, I think the `defaultLabelSelector` we pass to the Topology function (L308) should match the statefulset selector, otherwise it won't work: the `matchLabels` implements an "&&" condition and if the selector is different we have changes the topologySpreadConstraints is not properly applied during the scheduling.
In my environment I see:
```
}
{
  "matchLabels": {
    "app.kubernetes.io/name": "rabbitmq-cell1"
  }
}
{
  "matchLabels": {
    "app.kubernetes.io/name": "rabbitmq"
  }
}
```
so we should probably match only this one on L308, right?

Reply:
the docs here https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#components state the selector and pod template labels must be the same... but rabbitmqcluster doesn't do that

ultimately it doesn't really matter provided the selector includes the most specific label

Reply:
```
$ oc get pods rabbitmq-cell1-server-0 --show-labels
NAME                      READY   STATUS    RESTARTS   AGE   LABELS
rabbitmq-cell1-server-0   1/1     Running   0          15h   app.kubernetes.io/component=rabbitmq,app.kubernetes.io/name=rabbitmq-cell1,app.kubernetes.io/part-of=rabbitmq,apps.kubernetes.io/pod-index=0,controller-revision-hash=rabbitmq-cell1-server-55fdbd897d,statefulset.kubernetes.io/pod-name=rabbitmq-cell1-server-0

```